### PR TITLE
feature: Add gateway pattern (#1297)

### PR DESCRIPTION
**Pull request by FinnS-F**
Gateway Pattern

- Implemented gateway pattern described in issue https://github.com/iluwatar/java-design-patterns/issues/1297

### CODE REVIEW

File: gateway/README.md [hunk 1]
```diff
@@ -0,0 +1,151 @@
+---
+title: Gateway
+category: Structural
+language: en
+tag:
+
```
Comment (iluwatar) [inline gateway/README.md:6]: remove empty line

File: gateway/src/main/java/com/iluwatar/gateway/App.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    Gateway serviceB = gatewayFactory.getGateway("ServiceB");
+    Gateway serviceC = gatewayFactory.getGateway("ServiceC");
+
+    // Execute external services
+    try {
+      serviceA.execute();
+      serviceB.execute();
+      serviceC.execute();
+    } catch (ThreadDeath e) {
+      System.out.println("Interrupted!" + e);
```
Comment (iluwatar) [inline gateway/src/main/java/com/iluwatar/gateway/App.java:58]: Use a proper logger (Lombok's slf4j annotation works)

File: gateway/src/main/java/com/iluwatar/gateway/App.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+package com.iluwatar.gateway;
+
+/**
+ * The App class serves as the main entry point for the application implementing the Gateway design pattern.
+ * It demonstrates the use of the GatewayFactory to manage different gateway implementations, allowing for the
+ * execution of various services asynchronously.
+ *
+ * <p>In this example, there are three virtual service. GateFactory is the factory class and it provides a method
+ * to create different kinds of external services.
+ */
```
Comment (iluwatar) [inline gateway/src/main/java/com/iluwatar/gateway/App.java:34]: Explain briefly the pattern and how the example code implements it

File: gateway/src/main/java/com/iluwatar/gateway/ExternalServiceA.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+* ExternalServiceA is one of external services.
+*/
+class ExternalServiceA implements Gateway {
+  @Override
+  public void execute() throws Exception {
+    System.out.println("Executing Service A");
+    // Simulate a time-consuming task
+    Thread.sleep(1000);
+  }
+}
```
Comment (iluwatar) [inline gateway/src/main/java/com/iluwatar/gateway/ExternalServiceA.java:38]: Services A, B and C seem to be similar. Could we implement them with a single class?

**Review by iluwatar — CHANGES_REQUESTED** [review]

File: gateway/src/main/java/com/iluwatar/gateway/ExternalServiceA.java (hunk 4 above)
Comment (FinnS-F) [reply gateway/src/main/java/com/iluwatar/gateway/ExternalServiceA.java]: I have modified my code based on your review. For the last point you mentioned, these classes are just virtual implementations of external services. They may look the same(print out some string), but actually they are responsible for different functionalities.

File: gateway/README.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
+---
+title: Gateway
+category: Structural
+language: en
+tag:
+- Gang of Four
```
Comment (iluwatar) [inline gateway/README.md:6]: I think this is not included in the original GoF patterns

File: gateway/README.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+**Programmatic Example**
+
+The main class in our example is the `ExternalService` that contains items.
+
+```java
+class ExternalServiceA implements Gateway {
+    @Override
+    public void execute() throws Exception {
+        System.out.println("Executing Service A");
```
Comment (iluwatar) [inline gateway/README.md:41]: These should be updated to logger instances

**Review by iluwatar — APPROVED** [review]
Couple of minor comments, but I'll merge

**Comment by iluwatar** [comment]
@all-contributors please add @FinnS-F for code

**Comment by iluwatar** [comment]
https://github.com/iluwatar/java-design-patterns/pull/2818

**Pull request merged**